### DISCUSSION TITLE
Total elapsed time

### DISCUSSION THREAD

**Discussion by gudeh** [The-OpenROAD-Project/OpenROAD-flow-scripts#1660] [Q&A]
Hi everyone. 

On the end of log files there is usually something like:
5_3_route.log: `Elapsed time: 28:06.49[h:]min:sec. CPU time: user 6924.09 sys 34.88 (412%). Peak memory: 3802268KB.`
6_report.log:    `Elapsed time: 1:05.10[h:]min:sec. CPU time: user 53.04 sys 0.37 (82%). Peak memory: 619112KB.`

These elapsed times do not accumulate with each other, right? They are just the elapsed time of each individual step in the flow. I could not find the total elapsed time, is it available anywhere? If not I guess I can read each log file and sum the elapsed time values.

Also, this means that 5_3_route took 28minutes and 6.49 seconds, right?

**Comment by maliberty** [comment]
Yes the times are for each step of the value.  At the end we generate a summary table like
```
Log                       Elapsed seconds
1_1_yosys                        131
2_1_floorplan                     12
2_2_floorplan_io                   2
2_3_floorplan_tdms                 2
2_4_floorplan_macro                2
2_5_floorplan_tapcell              2
2_6_floorplan_pdn                  2
3_1_place_gp_skip_io              11
3_2_place_iop                      2
3_3_place_gp                      93
3_4_place_resized                 27
3_5_place_dp                      15
4_1_cts                          146
5_1_grt                          197
5_2_fillcell                       2
5_3_route                        286
6_1_merge                          5
6_report                          96
```

**Reply by gudeh** [reply]
Indeed, I see that on the terminal after the execution. Although, If I did not wrote that to a file, it is not saved.

**Reply by maliberty** [reply]
True but as the values are in the log files you can do 'make elapsed' and re-generate it very quickly.

**Comment by gudeh** [comment]
I did the following code to retrieve the elapsed time of the log files in each design and sum each one to get the total runtime of each design. Maybe this is useful for someone else also.

```
import os
import re

def extract_elapsed_time(file_path):
    with open(file_path, 'r') as file:
        content = file.read()
        match = re.search(r'Elapsed time: (\d+:\d{1,2}(:\d{1,2})?\.\d{2})', content)
        if match:
            return match.group(1)
        else:
            return None

def convert_to_seconds(elapsed_time):
    parts = elapsed_time.split(':')
    hours = int(parts[0])
    minutes, seconds = map(float, parts[1].split('.'))
    if len(parts) == 3:
        seconds += int(parts[2])
    total_seconds = hours * 3600 + minutes * 60 + seconds
    return total_seconds

def process_log_files(folder_path):
    elapsed_times = {}
    files = [f for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]
    
    for file_name in files:
        if file_name[0].isdigit() and file_name.endswith(".log"):
            file_path = os.path.join(folder_path, file_name)
            elapsed_time = extract_elapsed_time(file_path)
            if elapsed_time:
                elapsed_times[file_name] = elapsed_time
    
    return elapsed_times

def sum_elapsed_times(elapsed_times):
    total_seconds = 0
    for elapsed_time in elapsed_times.values():
        total_seconds += convert_to_seconds(elapsed_time)
    return total_seconds

def process_folders(root_folder='./'):
    for root, dirs, files in os.walk(root_folder):
        rel_path = os.path.relpath(root, root_folder)
        
        if os.path.basename(rel_path) == 'base':
            print(f"\nProcessing folder: {root}")
            result = process_log_files(root)
            for file_name, elapsed_time in result.items():
                print(f"{file_name}: Elapsed Time - {elapsed_time}")

            total_seconds = sum_elapsed_times(result)
            total_hours = total_seconds // 3600
            total_minutes = (total_seconds % 3600) // 60
            total_seconds_remainder = total_seconds % 60

            print(f"Total Elapsed Time: {int(total_hours)}:{int(total_minutes):02d}.{total_seconds_remainder:.2f}")

root_folder = './'
process_folders(root_folder)

```

**Reply by maliberty** [reply]
You could enhance https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/util/genElapsedTime.py which is what generates the above table.  It wouldn't be hard to add an overall sum.

**Reply by gudeh** [reply]
I did the following on branch "patch-2", on my own fork: https://github.com/gudeh/OpenROAD-flow-scripts/commit/7808ef106977705d68080dda12b451241e035cad

From the documentation, it seems I have to update my master branch on my fork prior to call the PR. Also, I see a failed test but I assume it is expected to fail. I am not sure if it is expected of me to change the test also.

**Reply by maliberty** [reply]
Yes you should update the test results to match your change.  Your branch should be reasonably up to date before opening the PR.
